### PR TITLE
Better Window.begin

### LINKED ISSUE #33: Weird Window.begin function
Is it better that I make it look like `win.begin()`?

### PR DESCRIPTION
**Pull request by iacore**
closes #33

### CODE REVIEW

File: src/dvui.zig [hunk 1]
```diff
@@ -2437,7 +2441,10 @@ pub const Window = struct {
             self.debug_under_mouse_info = "";
         }
 
+        _ = self._arena.reset(.retain_capacity);
```
Comment (iacore) [inline src/dvui.zig:2444]: previously you used `.free_all`. that's not how arena allocator should be used.

I also moved this call to the start of the frame.

File: src/dvui.zig [hunk 2]
```diff
@@ -2405,7 +2410,6 @@ pub const Window = struct {
 
     pub fn begin(
         self: *Self,
-        arena: std.mem.Allocator,
         time_ns: i128,
```
Comment (iacore) [inline src/dvui.zig:2413]: if possible, I think removing this argument is better.

**Comment by david-vanderson** [comment]
I like most of this, but I'd like to keep the begin/beginWait separation.  Can you back out that part?

**Comment by david-vanderson** [comment]
The idea is that if the application was already managing the framerate (say a game), then you only call Window.begin()/end().  If instead you want dvui to manage the framerate for you (normal application), then you add calls Window.beginWait(). and Window.waitTime().

It might not work out that way, but for now I'd like to keep that strategy.  I can merge and do that if you'd like?

**Comment by iacore** [comment]
Also you might want to merge the 2 SDL examples. It's better for maintenance.

File: sdl-test.zig [hunk 3]
```diff
@@ -34,14 +34,8 @@ pub fn main() !void {
...
 
-    var arena_allocator = std.heap.ArenaAllocator.init(std.heap.page_allocator);
-    const arena = arena_allocator.allocator();
-
     main_loop: while (true) {
-        defer _ = arena_allocator.reset(.free_all);
-
-        var nstime = win.beginWait(win_backend.hasEvent());
-        try win.begin(arena, nstime);
+        try win.begin(.{ .has_event = win_backend.hasEvent() });
```
[the remark this answers is not in the document]
Comment (david-vanderson) [reply sdl-test.zig:38]: My general rule of thumb for passing options into functions is whether a program will generally use a function with and without the option.  In this case, a given program will either use begin()/end() alone, or always add beginWait()/waitTime().  So I think it's clearer to have the separate functions.

I very much appreciate the push though!  API design is very difficult, and the only way to make it better is have a lot of people using it.  I agree with you that the original plan for the arena allocator wasn't great, and that definitely improves the api.  For all of these things, we can always revisit them if real-world use suggests things could be better.  Thank you!

[the remark this answers is not in the document]
Comment (iacore) [reply sdl-test.zig:38]: @david-vanderson is there a better way to contact you? Like Matrix or something. Github issue is not great for long discussions.

**Comment by david-vanderson** [comment]
For the backend idea, I'm trying to keep the option available to add dvui onto an existing application for debugging purposes.  It doesn't have much use yet, so probably things could be better, but the idea is that a game would already have a backend and pass it to dvui.  This was the idea behind the ontop-sdl.zig example - to show how to use dvui in a way that doesn't manage the backend for you.

That's also why there are two examples (ontop-sdl.zig and standalone-sdl.zig).  Does that make sense?  Or are you talking about something else?

**Comment by david-vanderson** [comment]
Should have said - the vsync thing is a reasonable concern.  I'm not sure the best way to make the distinction between how you get a backend in a standalone app vs. ontop situation.  Right now for standalone you use SDLBackend.init() which has the vsync option, and in an ontop situation you construct an SDLBackend manually:
`var backend = SDLBackend{ .window = window, .renderer = renderer };`

Does that make sense?

**Comment by iacore** [comment]
> For the backend idea, I'm trying to keep the option available to add dvui onto an existing application for debugging purposes. It doesn't have much use yet, so probably things could be better, but the idea is that a game would already have a backend and pass it to dvui. This was the idea behind the ontop-sdl.zig example - to show how to use dvui in a way that doesn't manage the backend for you.
> 
> That's also why there are two examples (ontop-sdl.zig and standalone-sdl.zig). Does that make sense? Or are you talking about something else?

The two examples share most of the code, so maybe making them as a single example with options to pick how to initialize backend would be nice.

**Comment by david-vanderson** [comment]
It's a fair point, but I'm trying to make it easy to start with the example that best fits the use-case someone is coming in with.  Again not sure that the ontop is the best, but I'd like to keep the separation for now.

Are you okay with me merging most of this?

**Comment by iacore** [comment]
> Are you okay with me merging most of this?

Go ahead!

**Comment by david-vanderson** [comment]
Thank you!

**Pull request merged**